src/gfn/estimators.py

```diff
@@ -798,3 +798,109 @@ def expected_output_dim(self) -> Optional[int]:
+    """Discrete policy estimator for recurrent architectures with explicit carry.
+
+    Many sequence models (e.g., RNN/LSTM/GRU/Transformer in autoregressive mode)
+    maintain a recurrent hidden state ("carry") that must be threaded through
+    successive calls during sampling. This class formalizes that pattern for
+    GFlowNet policies by:
+
+    - Exposing a forward signature ``forward(states, carry) -> (logits, carry)``
+      so the policy can update and return the next carry at each step.
+    - Requiring an ``init_carry(batch_size, device)`` method to allocate the
+      initial hidden state for a rollout.
+    - Ensuring the per-step output (``logits`` over actions) is derived from the
+      latest token/time step while the internal model may process sequences.
+
+    Interaction with the sampler/adapters
+    -------------------------------------
+    The sampler uses a ``RecurrentEstimatorAdapter`` which calls this estimator
+    with the current carry, updates the carry on every step, and records
+    per-step artifacts. Non-recurrent estimators should use the default adapter
+    and the standard ``DiscretePolicyEstimator`` base class instead.
+
+    Notes
+    -----
+    - Forward is intended for on-policy generation; off-policy evaluation over
+      entire trajectories typically requires different batching and masking.
+    - ``init_carry`` is a hard requirement for compatibility with the recurrent
+      adapter.
+    """
+
+    def __init__(
+        self,
+        module: nn.Module,
+        n_actions: int,
+        preprocessor: Preprocessor | None = IdentityPreprocessor(
+            output_dim=None
+        ),  # Addressed in https://github.com/GFNOrg/torchgfn/pull/399.
+        is_backward: bool = False,
+    ):
+        """Initializes a RecurrentDiscretePolicyEstimator.
+
+        Args:
+            module: The neural network module to use.
+            n_actions: Total number of actions in the discrete environment.
+            preprocessor: Preprocessor object that transforms states to tensors.
+        """
+        if preprocessor is None:
+            preprocessor = IdentityPreprocessor(
+                output_dim=None
+            )  # Addressed in https://github.com/GFNOrg/torchgfn/pull/399.
+        super().__init__(
+            module=module,
+            n_actions=n_actions,
+            preprocessor=preprocessor,
+            is_backward=is_backward,
+        )
+
+    def forward(
+        self,
+        states: States,
+        carry: dict[str, torch.Tensor],
+    ) -> tuple[torch.Tensor, dict[str, torch.Tensor]]:
+        """Forward pass of the module.
+
+        Args:
+            states: The input states.
+            carry: The carry from the previous step.
+
+        Returns:
+            The output of the module, as a tensor of shape (*batch_shape, output_dim).
+        """
+        # TODO: Is this still true? NOTE: Can only be used for on-policy generation,
```

Comment:
It looks like it. This forward is for on-policy generation only. For off-policy evaluation over entire trajectories, you need a separate path that handles padding/masks and batches multiple timesteps.

Reply:
@saleml based on some conversations today - I'm going to look at changing the `ctx` design to use the trajectories container instead.